tapchannel/allocation.go

```diff
@@ -0,0 +1,485 @@
+package tapchannel
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/lightninglabs/taproot-assets/address"
+	"github.com/lightninglabs/taproot-assets/asset"
+	"github.com/lightninglabs/taproot-assets/commitment"
+	"github.com/lightninglabs/taproot-assets/fn"
+	"github.com/lightninglabs/taproot-assets/proof"
+	"github.com/lightninglabs/taproot-assets/tappsbt"
+	"github.com/lightninglabs/taproot-assets/tapscript"
+	"github.com/lightninglabs/taproot-assets/tapsend"
+	"github.com/lightningnetwork/lnd/input"
+)
+
+var (
+	// ErrMissingInputs is an error that is returned if no inputs were
+	// provided.
+	ErrMissingInputs = fmt.Errorf("no inputs provided")
+
+	// ErrMissingAllocations is an error that is returned if no allocations
+	// were provided.
+	ErrMissingAllocations = fmt.Errorf("no allocations provided")
+
+	// ErrInputOutputSumMismatch is an error that is returned if the sum of
+	// the input asset proofs does not match the sum of the output
+	// allocations.
+	ErrInputOutputSumMismatch = fmt.Errorf("input and output sum mismatch")
+
+	// ErrNormalAssetsOnly is an error that is returned if an allocation
+	// contains an asset that is not a normal asset (e.g. a collectible).
+	ErrNormalAssetsOnly = fmt.Errorf("only normal assets are supported")
+
+	// ErrCommitmentNotSet is an error that is returned if the output
+	// commitment is not set for an allocation.
+	ErrCommitmentNotSet = fmt.Errorf("output commitment not set")
+)
+
+// AllocationType is an enum that defines the different types of asset
+// allocations that can be created.
+type AllocationType uint8
+
+const (
+	// AllocationTypeNoAssets is the default allocation type that is used
+	// when the allocation type is not important or the allocation does not
+	// carry any assets.
+	AllocationTypeNoAssets AllocationType = 0
+
+	// CommitAllocationToLocal is an allocation type that is used for
+	// allocating assets to the local party.
+	CommitAllocationToLocal AllocationType = 1
+
+	// CommitAllocationToRemote is an allocation type that is used for
+	// allocating assets to the remote party.
+	CommitAllocationToRemote AllocationType = 2
+
+	// CommitAllocationHtlcIncoming is an allocation type that is used for
+	// allocating assets to an incoming HTLC output.
+	CommitAllocationHtlcIncoming AllocationType = 3
+
+	// CommitAllocationHtlcOutgoing is an allocation type that is used for
+	// allocating assets to an outgoing HTLC output.
+	CommitAllocationHtlcOutgoing AllocationType = 4
+
+	// SecondLevelHtlcAllocation is an allocation type that is used for
+	// allocating assets to a second level HTLC output.
```

Comment:
What does "second level" mean here? Maybe we should add more to this comment to explain.

Reply:
Extended the comment with the wording from the BOLTS: https://github.com/lightning/bolts/blob/master/03-transactions.md#commitment-transaction-outputs